File: app/src/main/java/ru/neosvet/vestnewage/list/ListItem.java
package ru.neosvet.vestnewage.list;

import android.util.Pair;

import java.io.Serializable;
import java.util.Iterator;

import ru.neosvet.utils.NeoList;

public class ListItem implements Serializable {
    private String title, des = null;
    private final NeoList<String> heads = new NeoList<>();
    private final NeoList<String> links = new NeoList<>();
    private boolean select = false, few = false;

    public ListItem(String title, String link) {
        this.title = title;
        addLink(link);
    }

    public ListItem(String title) {
        this.title = title;
    }

    public ListItem(String title, boolean onlyTitle) {
        this.title = title;
        if (onlyTitle)
            addLink("#");
    }

    public boolean hasLink() {
        return links.isNotEmpty();
    }

    public boolean hasFewLinks() {
        return few;
    }

    public void clear() {
        heads.clear();
        links.clear();
        few = false;
    }

    public String getTitle() {
        return title;
    }

    public void setTitle(String title) {
        this.title = title;
    }

    public String getDes() {
        if (des == null) return "";
        return des;
    }

    public void setDes(String des) {
        this.des = des;
    }

    public String getLink() {
        if (links.isNotEmpty())
            return links.first();
        else
            return "";
    }

    public void addLink(String head, String link) {
        heads.add(head);
        addLink(link);
    }

    public void addLink(String link) {
        if (!few)
            few = links.isNotEmpty();
        links.add(link);
    }

    public void addHead(String head) {
        heads.add(head);
    }

    public String getHead() {
        if (heads.isNotEmpty())
            return heads.first();
        else
            return "";
    }

    public boolean isSelect() {
        return select;
    }

    public Iterator<String> getLinks() {
        links.reset();
        return links;
    }

    public Iterator<Pair<String, String>> headsAndLinks() {
        heads.reset();
        links.reset();
        return new Iterator<>() {
            @Override
            public boolean hasNext() {
                return heads.hasNext() && links.hasNext();
            }

            @Override
            public Pair<String, String> next() {
                return new Pair<>(heads.next(), links.next());
            }
        };
    }
}

File: app/src/main/java/ru/neosvet/vestnewage/data/ListItem.kt
package ru.neosvet.vestnewage.data

import android.util.Pair

class ListItem {
    var title: String
    var des: String = ""
    private val heads = NeoList<String>()
    private val listLinks = NeoList<String>()
    private var few = false

    constructor(title: String, link: String) {
        this.title = title
        addLink(link)
    }

    constructor(title: String) {
        this.title = title
    }

    constructor(main: Array<String>) {
        title = main[0]
        if (main[1].isNotEmpty())
            addHead(main[1])
        addLink(main[2])
        des = main[3]
    }

    constructor(title: String, onlyTitle: Boolean) {
        this.title = title
        if (onlyTitle) addLink("#")
    }

    val link: String
        get() = if (listLinks.isNotEmpty) listLinks.first() else ""

    val head: String
        get() = if (heads.isNotEmpty) heads.first() else ""

    val links: Iterator<String>
        get() {
            listLinks.reset(true)
            return listLinks
        }

    val main: Array<String>
        get() = arrayOf(title, head, link, des)

    fun hasLink(): Boolean {
        return listLinks.isNotEmpty
    }

    fun hasFewLinks(): Boolean {
        return few
    }

    fun clear() {
        heads.clear()
        listLinks.clear()
        few = false
    }

    fun addLink(head: String, link: String) {
        heads.add(head)
        addLink(link)
    }

    fun addLink(link: String) {
        if (!few) few = listLinks.isNotEmpty
        listLinks.add(link)
    }

    fun addHead(head: String) {
        heads.add(head)
    }

    fun headsAndLinks(): Iterator<Pair<String, String>> {
        heads.reset(true)
        listLinks.reset(true)
        return object : Iterator<Pair<String, String>> {
            override fun hasNext(): Boolean {
                return heads.hasNext() && listLinks.hasNext()
            }

            override fun next(): Pair<String, String> {
                return Pair(heads.next(), listLinks.next())
            }
        }
    }

    override fun toString(): String {
        return if (few)
            this.javaClass.simpleName + "[title=$title, links=$listLinks, des=$des, heads=$heads]"
        else
            this.javaClass.simpleName + "[title=$title, link=$link, des=$des, head=$head]"
    }
}